### ISSUE TITLE
Improve handling of platform-specific text labels in the UI

### ISSUE THREAD

**Issue by j-f1** [desktop/desktop#2103]
This continues my work in #2092.

Currently, the vast majority of the text of menu items are specified using a process similar to this:

```ts
__DARWIN__ ? 'New Repository…' : 'New &repository…'
```

This is suboptimal for several reasons:
1. The text is duplicated. If someone decides to change one label to make it clearer, they could forget to change the one for the other platform.
1. The titles are almost always quite similar. It would be fairly simple to automatically generate the macOS title from the Windows title, because all it needs is title-casing and the removal of `&` characters.
1. There is no way to enforce a consistent style guide for menu titles, because they’re scattered throughout various files, and there are similar lines like this:
    ```ts
    case 'alt': return __DARWIN__ ? '⌥' : 'Alt'
    ```
    It would be quite difficult to ignore these while still locating all of the menu titles.
1. If GitHub decides to offer language support in the app, it would quickly become unwieldy to choose the correct title for the correct language.

I propose this solution:

```ts
l('new repo') // `l` stands for “label.” This name could be changed to anything.
```
```yaml
# strings/en_US.yml
new repo: New &repository…
```

This resolves all of the issues above:
1. The text is not duplicated. A [custom TSLint plugin](https://palantir.github.io/tslint/develop/custom-rules/) could check any calls to `menuTitle` to verify that they exist in the strings file.
1. The title is automatically transformed from Windows style to macOS style.
1. A script could load the YAML file and verify that all of the items match the style guides.
1. Adding a new language is as simple as adding an appropriately-named file in the `strings` directory.

Issues raised in #2092 that are not solved with this proposal:

- [ ] How can we detect overlapping shortcuts with menu item names? [This comment](https://github.com/desktop/desktop/pull/2008#pullrequestreview-44573977) was the result of a human noticing it - are we able to automate these checks and not accidentally introduce a breakage?
  - This would seem to require parsing the TypeScript to determine which `l` calls are in the same menu, then locating those strings in the YAML file and checking the shortcuts.

**Comment by joshaber** [comment]
Yeah, this has been a known less-than-idealism for a while. One of the reasons we haven't tried to address it yet is that it's full of edge cases, and the actual cost of the problem is low while the cost of the solution is relatively high.

There are a few other things we need to take into account:

1. Some words or word combinations have complex capitalization (e.g., `GitHub`, `GitHub Enterprise`, `macOS`).
2. Sometimes the words themselves change based on the platform (e.g., [exit vs. quit](https://github.com/desktop/desktop/blob/1aa23acadcb7e8ea09eff9fd27e9ff2f713e77bb/app/src/crash/crash-app.tsx#L166)).

As @shiftkey said, this isn't a high priority right now, but it's good to have an issue tracking it!

**Comment by j-f1** [comment]
1. I can add a transform after calling `titleCase` with a list of special words.

    ```ts
    const escapeForRe = /[-\/\\^$*+?.()|[\]{}]/g
    const words = new Map<RegExp, string>([
      'GitHub',
      'macOS',
    ].map(string => [string, new RegExp(string.replace(escapeForRe, '\\$&'), 'i')]))
    ```
2. I handled that with [this mapping](https://github.com/j-f1/forked-desktop/blob/7e244b4b2e7a9356cfa83a67076fbcf09e2c5991/app/src/lib/platform-support.ts#L14-L22).
